src/java/org/apache/cassandra/streaming/StreamSession.java

```diff
@@ -1338,18 +1339,26 @@ public String toString()
                '}';
     }
 
-    public StringBuilder boundStacktrace(StackTraceElement[] stacktrace, int limit)
+    public static void boundStackTrace(Throwable e, int limit, HashSet<Throwable> visited)
     {
-        if (limit >= stacktrace.length) {
-            throw new IllegalArgumentException("Invalid limit size: " + limit);
-        }
+        if (e == null)
+            return;
+
+        if (visited.contains(e))
+            throw new IllegalArgumentException("Exception cycle detected");
+        else
+            visited.add(e);
 
-        StringBuilder out = new StringBuilder(limit);
-        for (int i = 0; i < limit; i++)
+        if (e.getStackTrace().length == 0)
         {
-            out.append('\n');
-            out.append('\t').append(stacktrace[i]);
+            if (e != null)
+                boundStackTrace(e.getCause(), limit, visited);
+            return;
         }
-        return out;
+
+        StackTraceElement[] stackTrace = e.getStackTrace();
+        StackTraceElement[] limitedStackTrace = Arrays.copyOfRange(stackTrace, 0, limit);
+        e.setStackTrace(limitedStackTrace);
```

Comment:
you should not be changing the stack trace of exceptions in production code, this can have impact you are not aware of; it is not safe to make things mutable that were not planned to be mutable